torch/csrc/jit/passes/memory_planning.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
+
+namespace torch {
+namespace jit {
+enum class Strategy {
+  NAIVE = 0,
+};
+
+inline const char* toString(Strategy s) {
+  switch (s) {
+    case Strategy::NAIVE:
+      return "NAIVE";
+    default:
+      return "UNKNOWN STRATEGY";
+  }
+}
+
+inline std::ostream& operator<<(std::ostream& str, Strategy rhs) {
+  return str << toString(rhs);
+}
+
+typedef struct MemRegion {
+  size_t offset;
+  size_t size;
+} MemRegion;
+
+inline std::ostream& operator<<(std::ostream& str, MemRegion reg) {
+  return str << "{offset: " << reg.offset << ", size: " << reg.size << "}";
+}
+
+inline bool operator==(const MemRegion& lhs, const MemRegion& rhs) {
+  return lhs.offset == rhs.offset && lhs.size == rhs.size;
+}
+
+struct regionSizeCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.size == reg2.size ? reg1.offset < reg2.offset
+                                  : reg1.size < reg2.size;
+  }
+};
+
+struct regionOffsetCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.offset == reg2.offset ? reg1.size < reg2.size
+                                      : reg1.offset < reg2.offset;
+  }
+};
+
+bool overlapMemRegion(const MemRegion& reg1, const MemRegion& reg2);
+
+struct UniqueLiveRange {
+  LiveRange lvr;
+  std::string id;
+};
+
+bool overlapLiveRange(
+    const UniqueLiveRange& ulvr1,
+    const UniqueLiveRange& ulvr2);
+
+inline std::ostream& operator<<(std::ostream& str, UniqueLiveRange rhs) {
+  return str << "{id: " << rhs.id << ", lvr: " << rhs.lvr << "}";
+}
+
+inline bool operator==(const UniqueLiveRange lhs, const UniqueLiveRange rhs) {
+  return lhs.lvr == rhs.lvr && lhs.id == rhs.id;
+}
+
+struct liveRangeStartCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.begin == u2.lvr.begin
+        ? (u1.lvr.end == u2.lvr.end ? u1.id < u2.id : u1.lvr.end < u2.lvr.end)
+        : u1.lvr.begin < u2.lvr.begin;
+  }
+};
+
+struct liveRangeEndCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.end == u2.lvr.end
+        ? (u1.lvr.begin == u2.lvr.begin ? u1.id < u2.id
+                                        : u1.lvr.begin < u2.lvr.begin)
+        : u1.lvr.end < u2.lvr.end;
+  }
+};
+
+template <typename Value>
+using SortedLiveRangeMap = std::map<UniqueLiveRange, Value, liveRangeStartCmp>;
```

Comment:
nit: `Value` is an overloaded term in compilers, I would make this `template <typename T>`

Reply:
not sure i understand - this basically comes from static runtime where the alias analysis creates a pairing between value and lifetimes

Reply:
I commented this before I saw the pr description my bad... but in-line comments good too

Reply:
sorry, I just mean `template <typename Value>` -> `template <typename T>

Reply:
right right